File: code/aiming_transporting/matlab/modules/plotting.m
function plot_trajectory(theta_val, tau_val, params)
% 绘制电子束轨迹
figure;
hold on;

% 计算轨迹点
t = linspace(0, tau_val, 100);
x = params.a0_max * sin(theta_val) * (1 - cos(params.wB * t));
y = params.a0_max * sin(theta_val) * sin(params.wB * t);
z = params.v0 * cos(theta_val) * t;

% 绘制3D轨迹
plot3(z*1e-3, x*1e-3, y*1e-3, 'LineWidth', 2);
plot3([0 params.R_m*cos(params.Psi)*1e-3], [0 0], [0 params.R_m*sin(params.Psi)*1e-3], 'r--');

% 标注目标点
plot3(params.R_m*cos(params.Psi)*1e-3, 0, params.R_m*sin(params.Psi)*1e-3, 'ro', 'MarkerSize', 10);

% 设置图形属性
xlabel('距离 (km)');
ylabel('横向位移 (km)');
zlabel('高度 (km)');
title(sprintf('电子束轨迹 (θ=%.2f rad)', theta_val));
grid on;
axis equal;
view(30, 30);
end

function plot_envelope(sol, params)
% 绘制束流包络
figure;
hold on;

% 获取解数据
z = sol.x;
R = sol.y(1,:);

% 绘制包络
plot(z*1e-3, R*1e2, 'b-', 'LineWidth', 2);
plot(z*1e-3, -R*1e2, 'b-', 'LineWidth', 2);

% 设置图形属性
xlabel('传播距离 (km)');
ylabel('束流半径 (cm)');
title('束流包络演化');
grid on;
end

File: code/aiming_transporting/modules/plotting.py
from modules.beam_params import BeamParams
import numpy as np
import matplotlib.pyplot as plt

def plot_trajectory(theta_val, tau_val, params: BeamParams):
    a0_val = params.a0_max * np.sin(theta_val)
    t = np.linspace(0, tau_val, 1000)
    x = a0_val * (1 - np.cos(params.wB * t))
    y = a0_val * np.sin(params.wB * t)
    z = params.v0 * np.cos(theta_val) * t
    xT = a0_val * (1 - np.cos(params.wB * tau_val))
    yT = a0_val * np.sin(params.wB * tau_val)
    zT = params.v0 * np.cos(theta_val) * tau_val
    phi = np.arctan2(-xT, yT)
    xp = np.cos(phi) * x + np.sin(phi) * y
    yp = -np.sin(phi) * x + np.cos(phi) * y
    xTp = np.cos(phi) * xT + np.sin(phi) * yT
    yTp = -np.sin(phi) * xT + np.cos(phi) * yT
    fig = plt.figure(figsize=(8,6))
    ax = fig.add_subplot(111, projection='3d')
    ax.plot(xp, yp, z)
    ax.plot([0, xTp], [0, yTp], [0, zT])
    ax.set_xlabel('x [m]')
    ax.set_ylabel('y [m]')
    ax.set_zlabel('z [m]')
    ax.set_title('电子束打击靶目标轨迹示意图')
    info = f'目标距离: {params.R_m*1e-3:.1f} km, 目标角度: {params.Psi:.2f} rad, 打击用时: {tau_val*1e6:.2f} $\\mu$ s'
    plt.annotate(info, xy=(0.05, 0.95), xycoords='axes fraction', fontsize=12, color='black', verticalalignment='top', bbox=dict(boxstyle='round', fc='w', alpha=0.7))
    plt.tight_layout()
    plt.savefig(f'aiming_R{(params.R_m)}_Psi{(params.Psi)}.jpg', dpi=300, bbox_inches='tight')
    plt.show()

def plot_envelope(sol, params: BeamParams):
    plt.plot(sol.t * 1e-3, sol.y[0] * 100)  # x轴: m->km, y轴: m->cm
    plt.xlabel('传输距离(km)')
    plt.ylabel('束流包络半径(cm)')
    plt.title('束流包络半径随传输距离的变化')
    info = f'目标距离: {params.R} km, 电子束电流: {params.Ib} A, 背景离子密度: {params.n_i:.2e} cm^-3'
    plt.annotate(info, xy=(0.05, 0.95), xycoords='axes fraction', fontsize=12, color='black', verticalalignment='top', bbox=dict(boxstyle='round', fc='w', alpha=0.7))
    plt.tight_layout()
    plt.savefig(f'envelop_R{(params.R)}_R_0{(params.R0)}.jpg', dpi=300, bbox_inches='tight')
    plt.grid()
    plt.show()